aten/src/ATen/core/function_schema.cpp

```diff
@@ -0,0 +1,142 @@
+#pragma once
+
+#include <ATen/core/function_schema.h>
+
+namespace c10 {
+
+std::ostream& operator<<(std::ostream& out, const FunctionSchema& schema) {
+  // eventually this should look almost identical to python arg parser, but
+  // it is simpler for now to work directly on this schema
+
+  out << schema.name();
+  out << "(";
+
+  bool seen_kwarg_only = false;
+  for(size_t i = 0; i < schema.arguments().size(); ++i) {
+    if (i > 0) out << ", ";
+    if (schema.arguments()[i].kwarg_only() && !seen_kwarg_only) {
+      out << "*, ";
+      seen_kwarg_only = true;
+    }
+    out << schema.arguments()[i];
+  }
+
+  if(schema.is_vararg()) {
+    if(schema.arguments().size() > 0)
+      out << ", ";
+    out << "...";
+  }
+
+  out << ") -> ";
+  if (schema.returns().size() == 1) {
+    out << schema.returns().at(0).type()->str();
+  } else if (schema.returns().size() > 1) {
+    out << "(";
+    for (size_t i = 0; i < schema.returns().size(); ++i) {
+      if (i > 0) out << ", ";
+      out << schema.returns()[i].type()->str();
+    }
+    out << ")";
+  }
+  return out;
+}
+
+void FunctionSchema::checkArg(const IValue& value, const Argument& argument, optional<size_t> pos) const {
+  if (!isSubvalueOf(value, argument.type())) {
+    std::string position = pos ? ::c10::str("in position ", *pos) : "";
+    AT_ERROR(
+        "Expected value of type ",
+        *argument.type(),
+        " for argument '",
+        argument.name(),
+        "'",
+        position,
+        ", but instead got value of type ",
+        attemptToRecoverType(value)->str(),
+        ". Declaration: ",
+        *this);
+  }
+}
+
+void FunctionSchema::findErrorInKwargs(const std::vector<std::string>& kwargs) const {
+  // First check if any of the kwargs are unknown, i.e. don't match the name of
+  // any argument in the schema.
+  for (const auto& kwarg : kwargs) {
```

Comment:
Why not `std::find` like in the loop below? Also, I wonder if it's worth putting both `arguments` and `kwargs` into an `unordered_set`.

Reply:
it needs to be a lambda variant because it is matching argument.name against kwarg. Not sure why it was count_if rather than find_if, I moved it from another file...

Reply:
Let's keep it as-is since it's just a move then. We can consider cleaning it up later.